modality-probe-graph/src/lib.rs

```diff
@@ -0,0 +1,617 @@
+                    self.graph.add_node(node);
+                    if first_event {
+                        if let Some(tail) =
+                            self.tail_pending_edge_sources.remove(&(probe_id, seq_num))
+                        {
+                            self.graph.add_edge(tail, node);
+                        }
+                        if let Some(tail) = self
+                            .last_event_by_probe_and_seq_num
+                            .remove(&(probe_id, seq_num))
+                        {
+                            self.graph.add_edge(tail, node);
+                        }
+                        first_event = false;
+                    }
+                    if let Some(prev) = prev_event {
+                        self.graph.add_edge(prev, node);
+                    }
+                    for lc in pending_edges.iter() {
+                        if let Some(e) = self.last_event_by_probe_and_clock.get(lc) {
+                            self.graph.add_edge(*e, node);
+                        }
+                    }
+                    pending_edges.clear();
+                    prev_event = Some(node);
+                }
+                EventLogEntry::EventWithPayload(id, payload) => {
+                    let node = GraphEvent {
+                        probe_id,
+                        id: *id,
+                        clock: *self_clock,
+                        payload: Some(*payload),
+                        seq: seq_num,
+                        seq_idx: idx,
+                    };
+                    self.graph.add_node(node);
+                    if first_event {
+                        if let Some(tail) =
+                            self.tail_pending_edge_sources.remove(&(probe_id, seq_num))
+                        {
+                            self.graph.add_edge(tail, node);
+                        }
+                        if let Some(tail) = self
+                            .last_event_by_probe_and_seq_num
+                            .remove(&(probe_id, seq_num))
+                        {
+                            self.graph.add_edge(tail, node);
+                        }
+                        first_event = false;
+                    }
+                    if let Some(prev) = prev_event {
+                        self.graph.add_edge(prev, node);
+                    }
+                    for lc in pending_edges.iter() {
+                        if let Some(e) = self.last_event_by_probe_and_clock.get(lc) {
+                            self.graph.add_edge(*e, node);
+                        }
+                    }
+                    pending_edges.clear();
+                    prev_event = Some(node);
+                }
+                EventLogEntry::TraceClock(lc) => {
+                    if lc.id == probe_id {
+                        // when we see a clock, the previous event we
+                        // saw is the last event from the previous
```

Comment:
Is this true even across Report boundaries when the Reports arrive out of order?

Reply:
It's not true across report boundaries, but this is all within a report boundary. This function just takes a single report "chunk" at a time.

Reply:
I'm concerned about the state consistency across reports, since this function deals with cross-report state stored in the `EventDigraph`. Here, I'm asking whether ` self.last_event_by_probe_and_clock` collection's consistency and assumed invariants hold when `EventDigraph::add_report` is called with reports that have arrived out of order.

Reply:
>  Here, I'm asking whether self.last_event_by_probe_and_clock collection's consistency and assumed invariants hold when EventDigraph::add_report is called with reports that have arrived out of order.

IIUYC, yes, they do. By

> but this is all within a report boundary

I meant that `prev_event` is only valid for a single call of `add_report`. It's initialized just before the loop, so this insert will only happen if there's an event from the previous iteration, within the current invocation of `add_report`. Therefore when we see a trace clock, but we've seen an event before **in the same report**, then it's safe to assume that event is the last event from the previous clock span. You're right in thinking that, across a report boundary, this could be inconsistent—in particular if the first entry in a report is a trace clock and `prev_event` had a dangling value. It's that `prev_event` is used within a single report that makes this work.

Reply:
Thinking this through a bit further, I think this does leave a hole when we can do better. If the first entry in a report is a self trace clock, we'll not try and lookup the last event from the previous sequence to add it to our `last_event_by_probe_and_clock`. Updating this now.